source/main.c

```diff
@@ -230,37 +243,48 @@ int main(void) {
         hsp = 0;
 
         // Player movement (pretty easy to understand)
-        // TODO: Would it be possible to make this less... iffy?
-        if (kHeld & KEY_UP) {
-            if (!(kHeld & KEY_DOWN)) {
-                vsp = -.5; // Vertical speed to negative .5
-                playerDir = BACK; // Player direction = back
-            }
-        }
-
-        if (kHeld & KEY_DOWN) {
-            vsp = .5; // Vertical speed to .5
-            playerDir = FORWARD; // Player direction = up
+        // TODO: Would it be possible to make this less... iffy? --- Sure. :P
+        //Thank GML for being VERY similar to C++ :)
+        //I'll explain my thought process clearly here.
+        //I'm going to seperate input, movement, and direction into different segments.
+        //  input
+        mkey_right = (kHeld & KEY_RIGHT); //assuming booleans work how I think they do here.. this should make mkey_right 1 when the right key is pressed.
```

Comment:
This is fine and all... but you aren't working with booleans right now. You're working with bit manipulation.

kDown is a 32 bit variable where each bit stands for a different button. KEY_DOWN, KEY_UP, etc., are bitmasks to select the appropriate bit, and only the appropriate bit. This ends up meaning that (kHeld & KEY_RIGHT) returns either 0, or KEY_RIGHT, since KEY_RIGHT (should) only select one bit. But KEY_RIGHT isn't 1... it's 1 << n.

Bit maniputation is very powerful and useful, but can be confusing. And if you're doing lots of it, well... you make lots of mistakes. There are a few ways to normalize this, but I'll take a closer look and see if there's something better that I can do later.